File: PDOO/QytetecJava/src/modeloqytetec/Calle.java

package modeloqytetec;

/**
 * @author Jorge Valenzuela Garcia
 */
public class Calle extends Casilla{
    
    // ATRIBUTOS
    private int numHoteles;
    private int numCasas;
    private TituloPropiedad titulo;
    
    // CONSTRUCTOR
    public Calle(int numeroCasilla, int coste, TituloPropiedad titulo){
      super(numeroCasilla, coste, TipoCasilla.CALLE);
      this.numHoteles = 0;
      this.numCasas = 0;
      this.titulo = titulo;
    }
    
    // METODOS
     public TituloPropiedad getTitulo(){
        return this.titulo;
    }
    
    protected TituloPropiedad asignarPropietario(Jugador jugador){
        this.titulo.setPropietario(jugador);
        return this.titulo;
    }
    
    protected int calcularValorHipoteca(){
        int hipobase = this.titulo.getHipotecaBase();
        int cantidadRecibida = hipobase + (int)((this.numCasas * 0.5 * hipobase) + (this.numHoteles * 2 * hipobase));
        return cantidadRecibida;
    }
            
    protected int cancelarHipoteca(){
        this.titulo.setHipotecada(false);
        int cantidadRecibida = this.calcularValorHipoteca();
        return cantidadRecibida;
    }
    
    
    protected int cobrarAlquiler(){
        int costeAlquilerBase = this.titulo.getAlquilerBase();
        int costeAlquiler = (int) (costeAlquilerBase + ((this.numCasas * 0.5) + (this.numHoteles * 2)));
        this.titulo.cobrarAlquiler(costeAlquiler);
        return costeAlquiler;
    }
            
    protected int edificarCasa(){
        this.setNumCasas(this.numCasas +1);
        return this.titulo.getPrecioEdificar();
    }
    
    protected int edificarHotel(){
        this.setNumHoteles(this.numHoteles +1);
        return this.titulo.getPrecioEdificar();
    }
    
    protected boolean estaHipotecada(){
        if(this.titulo.getHipotecada())
            return true;
        else
            return false;
    }
    
    protected int getNumCasas(){
        return this.numCasas;
    }
    
    protected int getNumHoteles(){
        return this.numHoteles;
    }
    
    protected int getPrecioEdificar(){
        return this.titulo.getPrecioEdificar();
    }
    
    protected int hipotecar(){
        this.titulo.setHipotecada(true);
        int cantidadRecibida = this.calcularValorHipoteca();
        return cantidadRecibida;
    }
    
    protected boolean propietarioEncarcelado(){
        return this.titulo.propietarioEncarcelado();
    }
    
    protected boolean tengoPropietario(){
        return this.titulo.tengoPropietario();
    }
    
    protected boolean sePuedeEdificarCasa(int factorEspeculador){
        return this.numCasas < 4*factorEspeculador;
    }
    
    protected boolean sePuedeEdificarHotel(int factorEspeculador){
        return this.numHoteles < 4*factorEspeculador && this.numCasas == 4*factorEspeculador;
    }
    
    protected void setNumCasas(int nuevoNumero){
        this.numCasas = nuevoNumero;
    }
    
    protected void setNumHoteles(int nuevoNumero){
        this.numHoteles = nuevoNumero;
    }
 
    private void setTitulo(TituloPropiedad titulo) throws Exception{
        this.titulo = titulo;
    }
    
    @Override
    protected boolean soyEdificable(){
        return true;
    }
    
    protected int venderTitulo(){
        this.titulo.setPropietario(null);
        this.setNumCasas(0);
        this.setNumHoteles(0);
        int precioVenta, precioCompra;
        precioCompra = super.getCoste() + ((this.numCasas + this.numHoteles) * this.titulo.getPrecioEdificar());
        precioVenta = (int) (precioCompra + (this.titulo.getFactorRevalorizacion() * precioCompra));
        return precioVenta;
    }
    
    
    
    @Override
    public String toString(){
        String salida;
        salida = "Numero Casilla: " + Integer.toString(this.getNumeroCasilla()) + "\n" +
        "Coste: " + Integer.toString(super.getCoste()) + "\n" +
        "Numero Hoteles: " + Integer.toString(this.numHoteles) + "\n" +
        "Numero Casas: " + Integer.toString(this.numCasas) + "\n" +
        "Tipo: CALLE\n";
        salida += "Titulo: " + this.titulo.toString();
        return salida;
    }
    
}

File: PDOO/QytetetRuby/lib/calle.rb



module ModeloQytetet
  class Calle < Casilla
  
  attr_accessor :num_hoteles, :num_casas, :titulo
  
  def initialize(numero_casilla, cost, tituloP)
    super(numero_casilla,TipoCasilla::CALLE ,cost)
    
    @titulo = tituloP
    @num_hoteles = 0
    @num_casas = 0
    
    if @titulo != nil
      @titulo.casilla = self
    end
  end
  
    
  def calcular_valor_hipoteca
    
      hipoteca_base = @titulo.hipoteca_base
      
       cantidad_recibida = hipoteca_base + (@num_casas * 0.5 * hipoteca_base + @num_hoteles * hipoteca_base)
      
      return cantidad_recibida
    end
    
    def cancelar_hipoteca
      
      @titulo.hipotecada = false
      cantidad_pagar = calcular_valor_hipoteca * 1.1
      
      return cantidad_pagar
    end
    
    def cobrar_alquiler
      
      coste_alquiler_base = @titulo.alquiler_base
      coste_alquiler = (coste_alquiler_base + (@num_casas * 0.5 + @num_hoteles * 2))
      @titulo.cobrar_alquiler(coste_alquiler)
      
      return coste_alquiler
    end
    
    def edificar_casa
      
      nuevo_num = @num_casas + 1
      @num_casas = nuevo_num
      coste_edificar_casas = @titulo.precio_edificar
      
      return coste_edificar_casas
    end
    
    def edificar_hotel
      nuevo_num = @num_hoteles + 1
      @num_hoteles = nuevo_num
      @num_casas = 0
      coste_edificar_hotel = @titulo.precio_edificar
      
      return coste_edificar_hotel
    end
    
    def esta_hipotecada
      return @titulo.hipotecada
    end
    
    
    def get_coste_hipoteca
      return calcular_valor_hipoteca
    end
    
   
    def get_precio_edificar
      coste_edificar_casa = @titulo.precio_edificar
      
      return coste_edificar_casa
    end
    
    def hipotecar
      @titulo.hipotecada = true
      cantidad_recibida = calcular_valor_hipoteca
      
      return cantidad_recibida
    end
    
    def precio_total_comprar
      
    end
    
    def propietario_encarcelado
      encarcelado = @titulo.propietario_encarcelado
      
      return encarcelado
    end
    
    def se_puede_edificar_casa(factor)
      return @num_casas < 4 * factor
    end
    
    def se_puede_edificar_hotel(factor)
      return @num_hoteles < (4 * factor) && @num_casas == 4 * factor
    end
    
     def tengo_propietario
      tengo_propietario = @titulo.tengo_propietario
      
      return tengo_propietario
    end
    
    def vender_titulo
      precio_compra = @coste + (@num_casas + @num_hoteles) * @titulo.precio_edificar
      precio_venta = precio_compra + @titulo.factor_revalorizacion * precio_compra
      
      @titulo.propietario = nil
      @num_hoteles = 0
      @num_casas = 0
      
      return precio_venta
    end
    
    def asignar_propietario(jugador)
      @titulo.propietario = jugador
      
      return @titulo
    end
    
    def asignar_titulo_propiedad
      
    end
    
    
    def to_s
      if @titulo.nil?
        puts "NumeroCasilla: #{@numero_casilla} \n Tipo: #{@tipo}"
      else
        puts "NumeroCasilla: #{@numero_casilla} \n Coste: #{@coste} \n NumHoteles #{@num_hoteles} \n NumCasas: #{@num_casas} \n Tipo: #{@tipo} \n Titulo:"
        puts @titulo.to_s
      end
    end

    
end
end
